### PR TITLE
Use dash generate components

### LINKED ISSUE #34: Build-dist script causes errors
As noted in https://github.com/plotly/dash-daq/pull/33, changing `build-dist-dev` to `build-dist` before publishing the NPM package results in an `Error loading dependencies`: 

screen shot 2019-02-18 at 1 47 36 pm

### LINKED ISSUE #61: Update generation to use `dash-generate-components`
Core component projects now use `dash-generate-components` from the dash repo to generate the Python and R wrappers.

### PR DESCRIPTION
**Pull request by shammamah**
Closes #61.
Closes #34.

## About 
* 🙃 
* Removed outdated `dash-components-archetype` in build process
* Added `dash-generate-components` 
* Updated webpack version
* Updated webpack config and bundle naming to be similar to other Dash core repos
* Removed `builder` dependency and all related files/scripts 
* Made bundle smaller by only including necessary files and the minified bundle
* Updated `styled-components` version to `4.x` (previous production versions were breaking, due to a warning in development mode turning into a breaking error in production mode: see https://github.com/styled-components/styled-components/issues/1709)
* Fixed enzyme/jest tests that were passing up to the last version of `styled-components` `3.x`, and failed upon update to `4.0.0` (my hypothesis is that this updated added another wrapping "layer" around components): 
  * Performed full mount instead of shallow mount in some cases, since child components had inaccessible props (see: https://airbnb.io/enzyme/docs/api/shallow.html)
  * Added in a call to `render()` in some cases (see: https://airbnb.io/enzyme/docs/api/ShallowWrapper/render.html), since `setProps` calls were not actually updating the props 
  * Changed checks of `{component}.state(...)` to `{component}.prop(...)`, since the state wasn't updating in tests (and it doesn't make a lot of sense to use `setProps` and then... not check the props
  * Fixed some errors/warnings in components related to outdated React lifecycle methods (e.g., `componentWillReceiveProps` -> `UNSAFE_componentWillReceiveProps` - see https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops); these will probably need to be changed later too 
  * Changed checks for number/value of `Digit` elements in the `PrecisionInput` component to also check for `ExponentialDigit`; even though `ExponentialDigit` "extends" `Digit`, the new version of `styled-components` gives them different classes so a `.find(Digit)` wasn't returning `ExponentialDigit` elements

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -2,11 +2,12 @@ build/
...
 dist/
 lib/
-lib/bundle.js*
+lib/*.js*
 node_modules/
 dash_daq/metadata.json
-dash_daq/bundle.js*
-inst/deps/bundle.js*
+dash_daq/*.js*
+inst/deps/*.js*
```
Comment (Marc-Andre-Rivet) [inline .eslintignore]: While I agree with the sentiment, we've been avoiding "all-inclusive" file selectors "just in case"

File: package.json [hunk 2]
```diff
@@ -59,13 +53,14 @@
     "react-color": "^2.14.1",
     "react-dom": "^16.8.6",
     "react-numeric-input": "^2.2.3",
-    "styled-components": "^3.1.6",
+    "styled-components": "^4.x",
```
Comment (alexcjohnson) [inline package.json]: Why `.x`? Why not `^4.4.0`?

File: package.json [hunk 3]
```diff
@@ -59,13 +53,14 @@
...
     "tinygradient": "^0.4.0",
-    "webpack": "^1.15.0",
+    "webpack": "^4.24.0",
```
Comment (alexcjohnson) [inline package.json]: That's a pretty big bump! But it's still a year old - current version is `4.41.0`.

If you're in the mode of updating deps, I recently updated dash-renderer to use babel@7 and corejs@3 - requires a little bit of tweaking (see https://github.com/plotly/dash/pull/903/files#diff-af8bab870558d6680886cec3032ab37c and other config file changes in that same PR) but it resulted in about a 20% drop in bundle size.

**Comment by Marc-Andre-Rivet** [comment]
`inst/deps` should not be removed except if you plan on making `dev` the default branch for DAQ, and the repo updated in a similar way as what was done in core. For now, suggest adding them back and make the master/dev artifact/no-artifact a follow up.

File: package.json (hunk 2 above)
Comment (shammamah) [reply package.json]: No particular reason - this was probably left over from when I was trying to figure out what was causing tests to break after `3.x`. Version has been specified in ff45419.

File: package.json (hunk 3 above)
Comment (shammamah) [reply package.json]: Deps updated in ff45419. It changed the bundle size from 566 kiB to 541 kiB!

File: package.json (hunk 2 above)
Comment (shammamah) [reply package.json]: No particular reason - this was probably left over from when I was trying to figure out what was causing tests to break after 3.x. Version has been specified in ff45419.

**Comment by shammamah** [comment]
I'll be regenerating the bundle/build artifacts once this PR gets approved (just in case the dependencies need to be changed).

File: package.json (hunk 3 above)
Comment (alexcjohnson) [reply package.json]: Nice! Not as big a change as we saw in renderer (253K -> 206K) but still worth the time I'd say 🎉

File: package.json [hunk 4]
```diff
@@ -13,44 +13,38 @@
...
     "prettier-restage": "git update-index --again"
   },
   "dependencies": {
-    "builder": "^3.2.2",
     "color": "^3.0.0",
     "conic-gradient": "^1.0.0",
     "copyfiles": "^1.2.0",
     "dash-components-archetype": "^0.2.11",
     "deep-equal": "^1.0.1",
+    "jest-canvas-mock": "^2.1.1",
```
Comment (alexcjohnson) [inline package.json]: devDependencies?

Comment (shammamah) [reply package.json]: Good call. Fixed in 9d1ce8a

File: package.json [hunk 5]
```diff
@@ -13,39 +13,32 @@
...
-    "test-frontend-cov": "jest --coverage --silent",
-    "test-watch": "builder run test-frontend-watch",
-    "test-debug": "builder run test-frontend-debug",
+    "start": "npm run build:dev",
+    "build:js": "webpack --mode production",
+    "build:js-dev": "webpack --mode development",
+    "build:py": "python get_version_info.py && yarn copy-lib && dash-generate-components ./src/components dash_daq -p package-info.json --r-prefix 'daq'",
+    "build": "npm run test && npm run lint && npm run build:js && npm run build:py",
+    "build:dev": "npm run build:js-dev && npm run build:py",
+    "build-tarball": "npm run build && python setup.py sdist",
```
Comment (alexcjohnson) [inline package.json]: you have a mix of `npm` and `yarn` here... probably not a problem but looks a little funny.

**Pull request merged**